raft/raft.go

```diff
@@ -0,0 +1,403 @@
+package raft
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/gatewayd-io/gatewayd/config"
+	pb "github.com/gatewayd-io/gatewayd/raft/proto"
+	"github.com/hashicorp/raft"
+	raftboltdb "github.com/hashicorp/raft-boltdb"
+	"github.com/rs/zerolog"
+	"google.golang.org/grpc"
+)
+
+// Command types for Raft operations.
+const (
+	CommandAddConsistentHashEntry = "ADD_CONSISTENT_HASH_ENTRY"
+	RaftLeaderState               = raft.Leader
+	LeaderElectionTimeout         = 3 * time.Second
+	maxSnapshots                  = 3                // Maximum number of snapshots to retain
+	maxPool                       = 3                // Maximum number of connections to pool
+	transportTimeout              = 10 * time.Second // Timeout for transport operations
+	leadershipCheckInterval       = 10 * time.Second // Interval for checking leadership status
+)
+
+// ConsistentHashCommand represents a command to modify the consistent hash.
+type ConsistentHashCommand struct {
+	Type      string `json:"type"`
+	Hash      uint64 `json:"hash"`
+	BlockName string `json:"blockName"`
+}
+
+// Node represents a node in the Raft cluster.
+type Node struct {
+	raft          *raft.Raft
+	config        *raft.Config
+	Fsm           *FSM
+	logStore      raft.LogStore
+	stableStore   raft.StableStore
+	snapshotStore raft.SnapshotStore
+	transport     raft.Transport
+	Logger        zerolog.Logger
+	Peers         []config.RaftPeer
+	rpcServer     *grpc.Server
+	rpcClient     *rpcClient
+	grpcAddr      string
+}
+
+// NewRaftNode creates and initializes a new Raft node.
+func NewRaftNode(logger zerolog.Logger, raftConfig config.Raft) (*Node, error) {
+	config := raft.DefaultConfig()
+
+	var err error
+	nodeID := raftConfig.NodeID
+	if raftConfig.NodeID == "" {
+		nodeID, err = os.Hostname()
+		if err != nil {
+			return nil, fmt.Errorf("error getting hostname: %w", err)
+		}
+	}
+	raftAddr := raftConfig.Address
+	config.LocalID = raft.ServerID(nodeID)
+	raftDir := filepath.Join(raftConfig.Directory, nodeID)
+	err = os.MkdirAll(raftDir, os.ModePerm)
+	if err != nil {
+		return nil, fmt.Errorf("error creating raft directory: %w", err)
+	}
+
+	// Create the FSM
+	fsm := NewFSM()
+
+	// Create the log store and stable store
+	logStore, err := raftboltdb.NewBoltStore(filepath.Join(raftDir, "raft-log.db"))
+	if err != nil {
+		return nil, fmt.Errorf("error creating log store: %w", err)
+	}
+
+	stableStore, err := raftboltdb.NewBoltStore(filepath.Join(raftDir, "raft-stable.db"))
+	if err != nil {
+		return nil, fmt.Errorf("error creating stable store: %w", err)
+	}
+
+	// Create the snapshot store
+	snapshotStore, err := raft.NewFileSnapshotStore(raftDir, maxSnapshots, os.Stderr)
+	if err != nil {
+		return nil, fmt.Errorf("error creating snapshot store: %w", err)
+	}
+
+	// Setup Raft communication
+	addr, err := net.ResolveTCPAddr("tcp", raftAddr)
```

Comment:
Does it only work with TCP?

Reply:
HashiCorp's Raft library primarily uses TCP for its default transport layer, as seen in the [raft.NewTCPTransport ](https://github.com/hashicorp/raft/blob/main/tcp_transport.go)function. However, the library is designed to be flexible and allows for custom transport implementations.